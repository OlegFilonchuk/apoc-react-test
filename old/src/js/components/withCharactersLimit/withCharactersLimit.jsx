import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './withCharactersLimit.less';

const withCharactersLimit = BaseComponent => {
  class DecoratedComponent extends Component {
    constructor(props) {
      super(props);

      this.getLettersLeftCount = this.getLettersLeftCount.bind(this);
      this.onChange = this.onChange.bind(this);
    }

    componentWillMount() {
      this.setValue(this.props);
    }

    componentWillReceiveProps(nextProps) {
      if (nextProps.value !== this.state.value) {
        this.setValue(nextProps);
      }
    }

    onChange(e) {
      e.persist();
      const value = e.target.value;

      this.setState({ value }, () => {
        this.props.onChange(e);
      });
    }

    setValue(props) {
      const value = props.value;

      this.setState(prevState => ({
        ...prevState,
        value
      }));
    }

    getLettersLeftCount(inputValue) {
      const { maxLength } = this.props;
      const offset = maxLength - inputValue.length;

      return offset < 0 ? 0 : offset;
    }

    render() {
      const { value } = this.state;
      const { onChange, className, ...restProps } = this.props;
      const lettersLeftCount = this.getLettersLeftCount(value);

      return (
        <div className="characters-limit">
          <BaseComponent
            {...restProps}
            onChange={this.onChange}
            value={value}
            className={`${className} with-characters-limit-input`}
          />
          <div className="input-field__char-limit-counter">
            <div className="input-field__char-limit-counter__text">{lettersLeftCount}</div>
          </div>
        </div>
      );
    }
  }

  DecoratedComponent.propTypes = {
    ...BaseComponent.propTypes,
    /**
     * Maximum length of the textarea field
     */
    maxLength: PropTypes.number,
    value: PropTypes.string
  };

  DecoratedComponent.defaultProps = {
    ...BaseComponent.defaultProps,
    maxLength: 280,
    value: ''
  };

  DecoratedComponent.displayName = `${BaseComponent.displayName}WithCharactersLimit`;

  return DecoratedComponent;
};

export default withCharactersLimit;
