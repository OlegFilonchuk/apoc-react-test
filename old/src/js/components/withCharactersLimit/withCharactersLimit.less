@counter-color: #555555;

.characters-limit {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  flex: 1;
  &__letters-left {
    color: @counter-color;
  }
}
