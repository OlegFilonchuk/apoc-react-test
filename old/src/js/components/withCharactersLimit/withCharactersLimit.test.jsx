import React from 'react';
import { mount } from 'enzyme';

import withCharactersLimit from './withCharactersLimit';

describe('withCharactersLimit HoC', () => {
  const BaseComponent = props => <input {...props} />;

  BaseComponent.displayName = 'BaseComponent';

  const mountWithCharactersLimit = (props, mountFn = mount) => {
    const ComponentWithCharacterLimit = withCharactersLimit(BaseComponent);

    return mountFn(<ComponentWithCharacterLimit {...props} />);
  };

  describe('render component', () => {
    it('should render div with characters limit', () => {
      const component = mountWithCharactersLimit({ value: 'test', maxLength: 10 });

      expect(component).toMatchSnapshot();
    });
  });

  describe('character limit', () => {
    it('should calculate proper left symbols count', () => {
      const component = mountWithCharactersLimit({ maxLength: 16, value: 'test' });
      const expectedCountBefore = '12';

      expect(component.find('.input-field__char-limit-counter__text').text()).toEqual(expectedCountBefore);

      const newValue = 'testtest';
      const expectedCountAfter = '8';

      component.setProps({ value: newValue });

      expect(component.find('.input-field__char-limit-counter__text').text()).toEqual(expectedCountAfter);
    });
  });
});
