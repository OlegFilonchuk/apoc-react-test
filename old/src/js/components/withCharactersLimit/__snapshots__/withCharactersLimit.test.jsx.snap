// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withCharactersLimit HoC render component should render div with characters limit 1`] = `
<BaseComponentWithCharactersLimit
  maxLength={10}
  value="test"
>
  <div
    className="characters-limit"
  >
    <BaseComponent
      className="undefined with-characters-limit-input"
      maxLength={10}
      onChange={[Function]}
      value="test"
    >
      <input
        className="undefined with-characters-limit-input"
        maxLength={10}
        onChange={[Function]}
        value="test"
      />
    </BaseComponent>
    <div
      className="input-field__char-limit-counter"
    >
      <div
        className="input-field__char-limit-counter__text"
      >
        6
      </div>
    </div>
  </div>
</BaseComponentWithCharactersLimit>
`;
